# According to the approch given in the Mail

In [1]:
pip install sentence-transformers

  Using cached nvidia_cuda_nvrtc_cu12-12.1.105-py3-none-manylinux1_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_cuda_runtime_cu12-12.1.105-py3-none-manylinux1_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_cuda_cupti_cu12-12.1.105-py3-none-manylinux1_x86_64.whl.metadata (1.6 kB)
  Using cached nvidia_cudnn_cu12-8.9.2.26-py3-none-manylinux1_x86_64.whl.metadata (1.6 kB)
  Using cached nvidia_cublas_cu12-12.1.3.1-py3-none-manylinux1_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_cufft_cu12-11.0.2.54-py3-none-manylinux1_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_curand_cu12-10.3.2.106-py3-none-manylinux1_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_cusolver_cu12-11.4.5.107-py3-none-manylinux1_x86_64.whl.metadata (1.6 kB)
  Using cached nvidia_cusparse_cu12-12.1.0.106-py3-none-manylinux1_x86_64.whl.metadata (1.6 kB)
  Using cached nvidia_nccl_cu12-2.20.5-py3-none-manylinux2014_x86_64.whl.metadata (1.8 kB)
  Using cached nvidia_nvtx_cu12-12.1.105-py3-none-manylinu

In [2]:
from sentence_transformers import SentenceTransformer,util

/usr/local/lib/python3.10/dist-packages/sentence_transformers/cross_encoder/CrossEncoder.py:11: TqdmExperimentalWarning: Using `tqdm.autonotebook.tqdm` in notebook mode. Use `tqdm.tqdm` instead to force console mode (e.g. in jupyter console)
  from tqdm.autonotebook import tqdm, trange


In [3]:
sentence1=["Cannot raise time-log request", "Can raise time-log request"]
sentence2=["Cannot raise time-log request", "Cannot raise service request"]

In [4]:
model=SentenceTransformer('sentence-transformers/all-MiniLM-L6-v2')
embedding1=model.encode(sentence1,convert_to_tensor=True)
embedding2=model.encode(sentence2,convert_to_tensor=True)

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.7k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

In [6]:
#print(embedding1)

In [ ]:
#cosine_similarity=util.cos_sim(embedding1,embedding2)
#print(cosine_similarity)

In [5]:
cosine_similarity_1=util.pytorch_cos_sim(embedding1[0],embedding1[1]).item()
cosine_similarity_2=util.pytorch_cos_sim(embedding2[0],embedding2[1]).item()

In [6]:
print(cosine_similarity_1)

0.8977203369140625


In [7]:
print(cosine_similarity_2)

0.5123158097267151


In [8]:
model.similarity(embedding1[0],embedding1[1]) # similar to cosine without uisng util

tensor([[0.8977]])

In [9]:
model.similarity(embedding2[0],embedding2[1]) # similar to cosine without uisng util

tensor([[0.5123]])

# Cross Checking with dot product score

In [12]:
dot_product1=util.dot_score(embedding1[0],embedding1[1]).item()
dot_product2=util.dot_score(embedding2[0],embedding2[1]).item()

In [13]:
print(dot_product1)

tensor([[0.8977]], device='cuda:0')


In [14]:
print(dot_product2)

tensor([[0.5123]], device='cuda:0')


# Creating a array of models and a loop to ruun. each model for cosine simailarty and dot product to save compuationals efforts and time

In [15]:
models=[
    'sentence-transformers/paraphrase-MiniLM-L6-v2',
    'sentence-transformers/all-mpnet-base-v2',
    'sentence-transformers/paraphrase-mpnet-base-v2',
    'sentence-transformers/all-distilroberta-v1',
    'sentence-transformers/all-roberta-large-v1',
    'sentence-transformers/paraphrase-distilroberta-base-v1',
    'sentence-transformers/nli-mpnet-base-v2',
    'sentence-transformers/nli-roberta-base-v2',
    'sentence-transformers/stsb-roberta-large',
    'sentence-transformers/all-MiniLM-L12-v2'
]

In [29]:
for modelname in models:
  model=SentenceTransformer(modelname)
  #print(modelname)
  print(f"\nModel: {modelname}")

  embedding1=model.encode(sentence1,convert_to_tensor=True)
  embedding2=model.encode(sentence2,convert_to_tensor=True)

  cosine_similarity_1=util.pytorch_cos_sim(embedding1[0],embedding1[1]).item()
  cosine_similarity_2=util.pytorch_cos_sim(embedding2[0],embedding2[1]).item()

  dot_product1=util.dot_score(embedding1[0],embedding1[1]).item()
  dot_product2=util.dot_score(embedding2[0],embedding2[1]).item()

  print(f"Cosine Similarity 1: {cosine_similarity_1}")
  print(f"Cosine Similarity 2: {cosine_similarity_2}")

  print(f"Dot Product 1: {dot_product1}")
  print(f"Dot Product 2: {dot_product2}")


Model: all-MiniLM-L6-v2
Cosine Similarity 1: 0.897720217704773
Cosine Similarity 2: 0.5123158693313599
Dot Product 1: 0.897720217704773
Dot Product 2: 0.5123158693313599

Model: paraphrase-MiniLM-L6-v2
Cosine Similarity 1: 0.8376305103302002
Cosine Similarity 2: 0.6131765842437744
Dot Product 1: 42.4522705078125
Dot Product 2: 34.03826141357422

Model: all-mpnet-base-v2
Cosine Similarity 1: 0.8442451357841492
Cosine Similarity 2: 0.5251204967498779
Dot Product 1: 0.8442450761795044
Dot Product 2: 0.5251203775405884

Model: paraphrase-mpnet-base-v2
Cosine Similarity 1: 0.9397901296615601
Cosine Similarity 2: 0.48978692293167114
Dot Product 1: 10.333602905273438
Dot Product 2: 5.364274501800537

Model: all-distilroberta-v1
Cosine Similarity 1: 0.8754093647003174
Cosine Similarity 2: 0.4455852508544922
Dot Product 1: 0.8754093647003174
Dot Product 2: 0.4455852508544922

Model: paraphrase-distilroberta-base-v1
Cosine Similarity 1: 0.8670663237571716
Cosine Similarity 2: 0.4945476651191711

# Only considering Cosine similarty and differnce , below is the updated code  for the same


In [31]:
for modelname in models:
  model=SentenceTransformer(modelname)
  #print(modelname)
  print(f"\nModel: {modelname}")

  embedding1=model.encode(sentence1,convert_to_tensor=True)
  embedding2=model.encode(sentence2,convert_to_tensor=True)

  cosine_similarity_1=util.pytorch_cos_sim(embedding1[0],embedding1[1]).item()
  cosine_similarity_2=util.pytorch_cos_sim(embedding2[0],embedding2[1]).item()

  #dot_product1=util.dot_score(embedding1[0],embedding1[1]).item()
  #dot_product2=util.dot_score(embedding2[0],embedding2[1]).item()

  print(f"Cosine Similarity 1: {cosine_similarity_1}")
  print(f"Cosine Similarity 2: {cosine_similarity_2}")
  print(f"\nCosine Similarity Difference: {abs(cosine_similarity_1-cosine_similarity_2)}")


Model: all-MiniLM-L6-v2
Cosine Similarity 1: 0.897720217704773
Cosine Similarity 2: 0.5123158693313599

Cosine Similarity Difference: 0.3854043483734131

Model: paraphrase-MiniLM-L6-v2
Cosine Similarity 1: 0.8376305103302002
Cosine Similarity 2: 0.6131765842437744

Cosine Similarity Difference: 0.22445392608642578

Model: all-mpnet-base-v2
Cosine Similarity 1: 0.8442451357841492
Cosine Similarity 2: 0.5251204967498779

Cosine Similarity Difference: 0.31912463903427124

Model: paraphrase-mpnet-base-v2
Cosine Similarity 1: 0.9397901296615601
Cosine Similarity 2: 0.48978692293167114

Cosine Similarity Difference: 0.4500032067298889

Model: all-distilroberta-v1
Cosine Similarity 1: 0.8754093647003174
Cosine Similarity 2: 0.4455852508544922

Cosine Similarity Difference: 0.4298241138458252

Model: paraphrase-distilroberta-base-v1
Cosine Similarity 1: 0.8670663237571716
Cosine Similarity 2: 0.49454766511917114

Cosine Similarity Difference: 0.3725186586380005

Model: all-roberta-large-v1
Co

In [ ]:
# higher differnce means more dissimilar

# Findings:
### The embedding which gave most difference was all-roberta-large-v1 where the differnce was of 0.4530905783176422.

### The cosine similarity of sentence 1 pair is 0.9397901 which contains "Cannot raise time-log request" and "Can raise time-log request.", this shows there is more similarity to the sentence even though there is differnce in the meaning in the context that is can and cannot

### The cosine similarity  of sentence 2 pair is  0.4726651608943939 , which indicate the pairs are not semanticly similar

In [ ]:
#####

# Some Observations:

### The all-roberta-large-v1 returned a high cosine score of 0.93 for the sentence pair having words can and cannot which was unexpected because these words should imply opposite meaning.

### On other hand nli-mpnet-base-v2 gave the cosine socre odf  0.6352491974830627which is more resonable score, recognizing the significant change in meaning caused by can and cannot words  and similarly for second sentence pair also which was appropiate for differnet meanings , but the cosine difference was 0.1  indicating higher similarty between both me senetence which is not true .

### On future reseraching i found out that the nli-mpnet model always handles these nuances better (can and cannot) but struggle in capturing context or meaning hence having 0.1 as the cosine differnce , where as the all-roberta-large-v1 capurtes the context quite well but sometimes fails in some nuances

### To tackel this we can use Negation-Aware Models or Rule-based Augmentation(can harm original meaning of text) or attentions mechnisms

# My approch( The way i intial thought to solve this problem) if the 'The Approch' was mentioned in the mail

In [32]:
pip install langchain langchain-community

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.4/50.4 kB 2.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 997.8/997.8 kB 18.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.3/2.3 MB 62.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 391.5/391.5 kB 36.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 140.4/140.4 kB 15.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.2/49.2 kB 5.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 141.9/141.9 kB 17.0 MB/s eta 0:00:00
  Attempting uninstall: tenacity
    Found existing installation: tenacity 9.0.0
    Uninstalling tenacity-9.0.0:
      Successfully uninstalled tenacity-9.0.0


In [33]:
%pip install --upgrade --quiet  langchain sentence_transformers

In [39]:
import getpass

inference_api_key = getpass.getpass("Enter your HF Inference API Key:\n\n")


Enter your HF Inference API Key:

··········


In [40]:
from langchain_community.embeddings import HuggingFaceInferenceAPIEmbeddings

In [41]:
embeddings = HuggingFaceInferenceAPIEmbeddings(api_key=inference_api_key, model_name="sentence-transformers/all-MiniLM-l6-v2")

In [42]:
sentence1=["Cannot raise time-log request", "Can raise time-log request"]
sentence2=["Cannot raise time-log request", "Cannot raise service request"]


In [44]:
embeddings1=embeddings.embed_documents(sentence1)
embeddings2=embeddings.embed_documents(sentence2)

In [45]:
import numpy as np

In [46]:
np.dot(embeddings1[0],embeddings1[1])

0.8977203073427534

In [47]:
np.dot(embeddings2[0],embeddings2[1])

0.5123156576481529

In [48]:
#print(np.dot(embeddings1[0],embeddings1[1])-np.dot(embeddings2[0],embeddings2[1]))
print(np.dot(embeddings1[0],embeddings1[1])-np.dot(embeddings2[0],embeddings2[1]))

0.38540464969460053
